Implementation of a better ranking to the grid

RamezIssac opened the pull request:
Hello 

Currently the DjangoPackages grid sorts packages by repo stars (named watchers in code).

This PR implements scoring a penalty of 10% of the stars for each 3 months the package is not updated; And score a penalty of -30% of the stars if it does not support python 3.
So an abandoned packaged for 2 years would lose 80% of its stars and would be moved backward in the grid.

If the main idea is accepted, we can move forward and fine tune those values, takes other parameters into account like status , etc.
Also check on how to initially update the ranking data.

This PR addresses #448 and possibly #565 

Kindest of Regards

RamezIssac commented:
Hello @vaibhav-jain 
Any updates on this ?

pydanny commented:
Sorry, we're all volunteers on this project, there are no sponsors. 😞 

In fact, I pay $60/month in hosting costs. Because of that, this has to be our lowest priority.

RamezIssac commented:
> Sorry, we're all volunteers on this project, there are no sponsors. 
> 
> In fact, I pay $60/month in hosting costs. Because of that, this has to be our lowest priority.

Totally understand, was just checking. :)

I take this opportunity to thank you , not just for DjangoPackages, but for many many important projects to the community.
Also, I'll contact you regarding a potential sponsoring of DjangoPackages.

Thank you again !

vaibhav-jain commented:
Hi @RamezIssac 

Could you add screenshots of how UI works after and before this PR?

RamezIssac commented:
Hello @vaibhav-jain 

I did not change anything related to the UI, i just added a new field called it rank and use it to order the packages instead of just using the stras (watchers in code)

Everything else stays the same.

> This PR implements scoring a penalty of 10% of the stars for each 3 months the package is not updated; And score a penalty of -30% of the stars if it does not support python 3.
So an abandoned packaged for 2 years would lose 80% of its stars and would be moved backward in the grid.

vaibhav-jain commented:
@RamezIssac Do you have working proof of this change? recording or something?

RamezIssac commented:
@vaibhav-jain How can i do that? 

The proof that this PR is working _should_ be the passing tests. (especially as i did not' change anything UI related).

Ii would love to push this forward but not sure exactly how :)
Give me database dump / fixtures to load to my local and runserver, send you a video for before and after ... for example ?

vaibhav-jain commented on package/tests/test_models.py:
```diff
@@ -7,24 +7,42 @@ class VersionTests(TestCase):
-                            '2.1.1',
-                            '2.1.2',
-                            '2.1.3']
+        expected_values = ['2.0.0',
```
fix this and then I'll be happy to test and merge this PR

RamezIssac replied:
This change was just a code format thing. will revert it and revise that only code change are included

RamezIssac commented:
Hello @vaibhav-jain , I removed any code formatting and rebased.

The PR in brief again:
It implements scoring a penalty of 10% of the stars for each 3 months the package is not updated; And score a penalty of -30% of the stars if it does not support python 3.
So an abandoned packaged for 2 years would lose 80% of its stars and would be moved backward in the grid.

What is not shipped with this PR is the step to calculate the rank for the existing packages... Can be added to the migration but i would need a confirmation from you first.

vaibhav-jain commented on package/tests/test_models.py:
```diff
@@ -7,6 +7,23 @@ class VersionTests(TestCase):
     def setUp(self):
         data.load()
 
+    def test_ranking(self):
```
@RamezIssac can we have more tests for a package that is last updated like 4 months ago, 9 months ago, 13 months ago 2.5 years ago?

I would suggest also add a test case to make sure your logic is working if the repo is archived or deleted from the GitHub

jefftriplett commented:
@RamezIssac thank you for the work on this. This is pretty high on my list of things to look at once I am a little bit more comfortable with the project. I might add tests in a new PR to this, but wanted to pass a long thank you comment until then.

RamezIssac commented:
@jefftriplett Thank you for this heart warming message :) I'm here for any possible further needed thing.
Best of luck :muscle:

jefftriplett commented:
I am hoping to get some time to look at this over the weekend. We merged in #683 which gives us a library we can use to overhaul our Python 3 detection with the trover classifier library (and other goodies). 

I also asked two data science friends for feedback and I need to write up some notes for a medium-term perspective. Short term, we should probably rename `rank` to `score` and then duplicate/add tests that test for a longer time period so that they are no surprises in grid scoring.

jefftriplett commented:
I ran into a few issues and started #700 so I can work with your branch. Hopefully, I can land this Monday or Tuesday this week.
